### PR TITLE
problem 53: Self-Attention Mechanism

### PR DESCRIPTION
**Pull request by Jayanth-vardhan**
This pull request introduces an implementation of the self-attention mechanism as part of problem 53. The changes include:

learn.html: This file contains a detailed explanation of the self-attention mechanism, its mathematical background, and practical implications.
solution.py: This file includes the Python implementation of the self-attention mechanism along with test cases to validate its functionality.

I look forward to your feedback and suggestions for improvements. Thank you for considering this addition!

### CODE REVIEW

File: Problems/53_self_attention/solution.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
+import numpy as np
+
+def self_attention(query, key, value):
```
Comment (moe18) [inline Problems/53_self_attention/solution.py:3]: I would like it to start from just given X and W and making the user compute Q, K, V and then use those values to compute attention

File: Problems/53_self_attention/learn.html [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    <meta name="viewport" content="width=device-width, initial-scale=1.0">
+    <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
+    <script id="MathJax-script" async src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
+    <title>Self-Attention Mechanism</title>
+</head>
+<body>
+    <h2>Self-Attention Mechanism</h2>
+
+    <p>This document provides an overview of the self-attention mechanism, which is fundamental in transformer models for tasks like natural language processing and computer vision.</p>
+
```
Comment (moe18) [inline Problems/53_self_attention/learn.html:14]: maybe put the Practical implementation here this way people know why attention is useful and how it works intuitively, but up to you

File: Problems/53_self_attention/learn.html [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+    <h3>Mathematical Background</h3>
+    <ul>
+        <li><strong>Self-Attention Calculation</strong>:
+            <br>
+            Given an input sequence \(X\):
+            \[
+            Q = XW^Q, \quad K = XW^K, \quad V = XW^V
```
Comment (moe18) [inline Problems/53_self_attention/learn.html:21]: I would change the ^Q, ^k ^V to a subscript something like this Q = XW_Q, \quad K = XW_K, \quad V = XW_V

**Review by moe18 — COMMENTED** [review]
it is looking really nice, added a few comments and suggestions

File: Problems/53_self_attention/solution.py (hunk 1 above)
Comment (Jayanth-vardhan) [reply Problems/53_self_attention/solution.py:3]: Thanks for the suggestion! I will modify the code to compute Q, K, and V based on the input X and W, and then proceed with the attention calculation.

File: Problems/53_self_attention/learn.html (hunk 3 above)
Comment (Jayanth-vardhan) [reply Problems/53_self_attention/learn.html:21]: Sure, I'll update the notation to use subscripts for clarity.

File: Problems/53_self_attention/learn.html (hunk 2 above)
Comment (Jayanth-vardhan) [reply Problems/53_self_attention/learn.html:14]: I’ll move the practical implementation earlier in the document to make it clearer why and how attention works.

**Comment by Jayanth-vardhan** [comment]
I've made the requested changes to both the HTML and Python files. Subscripts have been adjusted, and the practical implementation section has been updated in learn.html. Please review the updated version. Thank you!

**Review by moe18 — APPROVED** [review]

**Pull request merged**